### PR TITLE
clarify http retry attempts

### PR DESCRIPTION
**Pull request by linsun**
our default attempts is 2.  We didn't say it earlier because it depends on the value of http timeout.  Now timeout is disabled by default, the default retry attempts should be clear to our users.

### CODE REVIEW

File: networking/v1alpha3/virtual_service.proto [hunk 1]
```diff
@@ -1536,10 +1536,11 @@ message StringMatch {
...
 message HTTPRetry {
-  // Number of retries for a given request. The interval
-  // between retries will be determined automatically (25ms+). Actual
-  // number of retries attempted depends on the request `timeout` of the
-  // [HTTP route](https://istio.io/docs/reference/config/networking/virtual-service/#HTTPRoute).
+  // Number of retries for a given request, default is 2. The interval
+  // between retries will be determined automatically (25ms+). When request 
+  // `timeout` of the [HTTP route](https://istio.io/docs/reference/config/networking/virtual-service/#HTTPRoute) 
+  // is configured, the actual number of retries attempted also depends on
+  // the specified request `timeout` value.
```
Comment (ramaraochavali) [inline networking/v1alpha3/virtual_service.proto:1543]: and per-try-timeout also right?

**Review by hzxuzhonghu — APPROVED** [review]

File: networking/v1alpha3/virtual_service.proto (hunk 1 above)
Comment (linsun) [reply networking/v1alpha3/virtual_service.proto]: Updated, thanks @ramaraochavali

**Comment by linsun** [comment]
@nrjpoddar @rshriram @louiscryan could one of you approve this minor PR?

File: networking/v1alpha3/virtual_service.proto [hunk 2]
```diff
@@ -1536,10 +1536,11 @@ message StringMatch {
...
 message HTTPRetry {
-  // Number of retries for a given request. The interval
-  // between retries will be determined automatically (25ms+). Actual
-  // number of retries attempted depends on the request `timeout` of the
-  // [HTTP route](https://istio.io/docs/reference/config/networking/virtual-service/#HTTPRoute).
+  // Number of retries for a given request, default is 2. The interval
+  // between retries will be determined automatically (25ms+). When request 
+  // `timeout` of the [HTTP route](https://istio.io/docs/reference/config/networking/virtual-service/#HTTPRoute) 
+  // is configured, the actual number of retries attempted also depends on
+  // the specified request `timeout` and `per_try_timeout` values.
```
Comment (nrjpoddar) [inline networking/v1alpha3/virtual_service.proto:1543]: Is there a default value of `per_try_timeout`?

the actual number of retries attempted also depends on the specified request `timeout` and `per_try_timeout` values. ->  Should we explain how the number of retries are calculated?

Comment (linsun) [reply networking/v1alpha3/virtual_service.proto:1543]: default value is no timeout from [envoy doc](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-field-config-route-v3-hedgepolicy-hedge-on-per-try-timeout).  

```
If left unspecified, Envoy will use the global route timeout for the request. Consequently, when using a 5xx based retry policy, a request that times out will not be retried as the total timeout budget would have been exhausted.
```

**Comment by linsun** [comment]
@hzxuzhonghu @ramaraochavali @nrjpoddar @howardjohn  I need some help in determining the default num of retry attempts now.  It is not simple!  Given our http req timeout default is no timeout, and per_try_timeout is configured same as http req timeout when not specified (which istio doesn't specify by default)... I'm inclined to say we don't do actual retries... Is this conclusion crazy?

**Comment by ramaraochavali** [comment]
@linsun I think it is not correct to conclude there wont be any retries. If client is not able to connect to server i.e. Service Unavailable cases , retries do happen. If client is not able to connect, connect_timeout happens first (even before request_timeout) and request will be retried.  If it is able to connect and request is sent and under process then since we do not have a sensible/safe default for request timeouts - that request is not retried.

**Review by ramaraochavali — APPROVED** [review]

File: networking/v1alpha3/virtual_service.proto (hunk 2 above)
Comment (linsun) [reply networking/v1alpha3/virtual_service.proto:1543]: Regarding how retries are calculated, it is complicated. I think it is best to refer to envoy docs vs istio provides explanation here.

**Comment by linsun** [comment]
@ramaraochavali thanks for chime in.  I agree with your assessment.

**Pull request merged**